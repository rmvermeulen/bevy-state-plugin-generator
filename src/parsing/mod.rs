pub(crate) mod header;
mod parsers;
#[cfg(test)]
mod tests;
mod tokens;

pub use parsers::*;
pub(crate) use tokens::*;

// TODO: investigate "tree-less parsing", see https://www.youtube.com/watch?v=NxiKlnUtyio
