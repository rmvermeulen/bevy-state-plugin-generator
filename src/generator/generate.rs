use std::rc::Rc;

use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::PluginConfig;
use crate::config::PluginName;
use crate::generator::context::Context;
use crate::generator::state_defs::generate_all_state_definitions;
use crate::models::{DefinedStates, StateNode};
use crate::parsing::{ParseNode, parse_states_text};
use crate::tree::SubTree;

pub(super) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub trait ToStringWith {
    fn to_string_indented<S: AsRef<str>>(&self, join: S) -> String;
}

pub fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

pub(crate) fn generate_plugin_source_from_defined_states(
    defined_states: DefinedStates,
    config: PluginConfig,
) -> String {
    let PluginConfig {
        plugin_name,
        root_state_name,
        states_module_name,
        naming_scheme,
        additional_derives,
    } = config;

    let mut context = Context::from(naming_scheme);
    for derive in additional_derives {
        context.derives.push(derive.to_string());
    }

    let type_definitions = generate_all_state_definitions(defined_states, context);
    let definitions_source = type_definitions.to_string_indented("    ");

    let plugin_builder = if let Some(root_state_name) = root_state_name {
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = type_definitions
            .inner()
            .into_iter()
            .skip(1) // skip root
            .map(|typedef| typedef.typename)
            .map(|state_name| format!(".add_sub_state::<{states_module_name}::{state_name}>()"))
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states = type_definitions
            .inner()
            .into_iter()
            .map(|sdef| {
                if sdef.parent_name.is_some() {
                    let state_name = sdef.typename;
                    format!(".add_sub_state::<{states_module_name}::{state_name}>()")
                } else {
                    let state_name = sdef.typename;
                    format!(".init_state::<{states_module_name}::{state_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    formatdoc! {"
        #![allow(missing_docs)]
        use bevy::prelude::AppExtStates;
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "}
}

#[cfg(feature = "rustfmt")]
fn try_format_source(source: &str) -> std::io::Result<String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source)
        .stderr_to_stdout()
        .read()
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> String {
    let source = source.as_ref();
    #[cfg(feature = "rustfmt")]
    let source = try_format_source(source).unwrap_or_else(|_| source.to_owned());
    #[cfg(not(feature = "rustfmt"))]
    let source = source.to_owned();

    if source.ends_with(|c: char| c.is_newline()) {
        source
    } else {
        source + "\n"
    }
}

pub fn generate_state_plugin_source(
    source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, String> {
    let parse_nodes = parse_states_text(source).map_err(|e| e.to_string())?;
    let defined_states = if let Some(ref root_state_name) = plugin_config.root_state_name {
        let root_state_name = root_state_name.to_string();
        let parse_tree = if parse_nodes.is_empty() {
            ParseNode::singleton(root_state_name.as_str())
        } else {
            ParseNode::enumeration(root_state_name.as_str(), parse_nodes)
        };
        let parse_tree_size = parse_tree.get_tree_size();
        let root_node: Rc<StateNode> = parse_tree
            .try_into_state_node()
            .map(Rc::new)
            .map_err(|e| format!("{e:?}"))?;
        let state_tree_size = root_node.get_tree_size();

        if state_tree_size > parse_tree_size {
            return Err("state-tree exceeds parse-tree!".into());
        }

        DefinedStates::Root(root_node)
    } else {
        DefinedStates::Unrelated(
            parse_nodes
                .into_iter()
                .map(|node| node.try_into_state_node())
                .collect::<Result<Vec<_>, _>>()
                .unwrap(),
        )
    };

    let plugin_source = generate_plugin_source_from_defined_states(defined_states, plugin_config);
    let source = if let Some(src_path) = src_path {
        let debug_info = generate_debug_info(src_path, source);
        [debug_info, plugin_source].join("\n")
    } else {
        plugin_source
    };

    Ok(format_source(source))
}
