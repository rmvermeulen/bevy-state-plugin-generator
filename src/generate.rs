#[cfg(test)]
mod tests;

use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::PluginConfig;
use crate::parsing::parse_states_text;
use crate::processing::{ProcessingError, try_parse_node_into_final_source};

pub(super) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

#[cfg(feature = "rustfmt")]
fn try_format_source(source: &str) -> std::io::Result<String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source)
        .stderr_to_stdout()
        .read()
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> String {
    let source = source.as_ref();
    #[cfg(feature = "rustfmt")]
    let source = try_format_source(source).unwrap_or_else(|_| source.to_owned());
    #[cfg(not(feature = "rustfmt"))]
    let source = source.to_owned();

    if source.ends_with(|c: char| c.is_newline()) {
        source
    } else {
        source + "\n"
    }
}

pub fn generate_state_plugin_source(
    input_source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, ProcessingError> {
    let parse_nodes = parse_states_text(input_source)?;
    let mut output = parse_nodes
        .into_iter()
        .map(|parse_node| try_parse_node_into_final_source(parse_node, plugin_config.clone()))
        .collect::<Result<Vec<_>, _>>()?
        .join("\n");
    if let Some(src_path) = src_path {
        let debug_info = generate_debug_info(src_path, input_source);
        output = [debug_info, output].join("\n");
    }
    Ok(format_source(output))
}
