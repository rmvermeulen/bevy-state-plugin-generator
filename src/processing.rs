#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use bevy_platform::collections::HashSet;
use bevy_utils::default;
use indoc::formatdoc;
use itertools::{Itertools, concat};

use crate::generate::REQUIRED_DERIVES;
use crate::parsing::ParseNode;
use crate::{NamingScheme, PluginConfig, PluginName};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub enum NodeType {
    #[default]
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NodeData {
    pub index: usize,
    pub parent: Option<usize>,
    pub node_type: NodeType,
    pub depth: usize,
    pub name: String,
    pub resolved_name: Option<String>,
    pub variants: Vec<String>,
}

pub fn flatten_root_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                continue;
            }
        };
        let Some(name) = parse_node.name() else {
            continue;
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
            ..default()
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for i in 0..nodes.len() {
        let (ancestors, children) = nodes.split_at_mut(i);
        let node = &children[0];
        if let Some(parent_index) = node.parent {
            assert!(parent_index < i);
            let parent = &mut ancestors[parent_index];
            parent.variants.push(node.name.clone());
        }
    }

    #[cfg(debug_assertions)]
    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent_index) = node.parent {
            assert!(parent_index <= node.index);
        }
    }

    nodes
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    nodes: &mut [NodeData],
) -> Result<(), ProcessingError> {
    let mut names = HashSet::new();
    let mut resolved_names = Vec::new();
    for node in nodes.iter() {
        let base_name = &node.name;
        let resolved_name = match naming_scheme {
            NamingScheme::Full => {
                let mut ancestors = Vec::new();
                let mut current = node;
                while let Some(p) = current.parent {
                    current = &nodes[p];
                    ancestors.push(current);
                }
                let ancestral_name = ancestors.into_iter().map(|a| &a.name).rev().join("");
                format!("{ancestral_name}{base_name}")
            }
            NamingScheme::Short => {
                let parent = node
                    .parent
                    .map(|p| nodes[p].name.as_str())
                    .unwrap_or_default();
                format!("{parent}{base_name}")
            }
            NamingScheme::None => base_name.clone(),
        };
        if !names.insert(resolved_name.clone()) {
            return Err(ProcessingError::DuplicateName {
                resolved_name,
                original_name: base_name.clone(),
            });
        }

        resolved_names.push(resolved_name);
    }
    assert_eq!(names.len(), resolved_names.len());
    for (i, resolved_name) in resolved_names.into_iter().enumerate() {
        nodes[i].resolved_name = Some(resolved_name);
    }
    Ok(())
}

type NomError<T> = nom::Err<nom::error::Error<T>>;

#[derive(Debug, thiserror::Error)]
pub enum ProcessingError {
    #[error("Duplicate name: resolved_name='{resolved_name}' original_name='{original_name}'")]
    DuplicateName {
        resolved_name: String,
        original_name: String,
    },
    #[error("Unspecified error: {0}")]
    Custom(String),
    #[error("Failed to parse! Final state: {0:?}")]
    Parsing(#[from] nom::Err<nom::error::Error<String>>),
    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),
}

impl From<NomError<&str>> for ProcessingError {
    fn from(value: NomError<&str>) -> Self {
        Self::from(value.to_owned())
    }
}

fn build_plugin_source(
    nodes: Vec<NodeData>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let PluginConfig {
        additional_derives: derives,
        plugin_name,
        root_state_name,
        naming_scheme: _,
        states_module_name,
    } = config;

    let derives = concat([
        REQUIRED_DERIVES
            .iter()
            .map(ToString::to_string)
            .collect_vec(),
        derives.into_iter().collect_vec(),
    ])
    .into_iter()
    .unique()
    .join(", ");
    let definitions_source = nodes
        .iter()
        .map(|node| {
            let derives = node
                .parent
                .map(|parent_id| {
                    let parent = &nodes[parent_id];
                    formatdoc! {"
                        #[derive(bevy::prelude::SubStates, {derives})]
                        #[source({source} = {source}::{variant})]
                    ", variant = node.name, source = parent.name
                    }
                })
                .unwrap_or_else(|| formatdoc! {"#[derive(bevy::prelude::States, {derives})]"})
                .trim()
                .to_string();
            let typename = &node.name;

            let source_for_singleton = || {
                formatdoc! {"
                {derives}
                pub struct {typename};
            "}
            };
            let source_for_enum = |variants: &[&str]| {
                formatdoc! {"
                {derives}
                pub enum {typename} {{
                    #[default]
                    {variants}
                }}
                ", variants = variants.join(",\n      ")
                }
            };

            match node.node_type {
                NodeType::Singleton => source_for_singleton(),
                NodeType::Enum => {
                    if node.variants.is_empty() {
                        source_for_singleton()
                    } else {
                        let variants = node.variants.iter().map(String::as_str).collect_vec();
                        source_for_enum(&variants)
                    }
                }
                NodeType::List => source_for_singleton(),
            }
        })
        .join("\n");

    let plugin_builder = if let Some(root_state_name) = root_state_name.as_ref() {
        let states_module_name = states_module_name.as_str();
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = nodes
            .iter()
            .flat_map(|node| {
                let resolved_name = node
                    .resolved_name
                    .as_ref()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    Some(format!(
                        ".add_sub_state::<{states_module_name}::{resolved_name}>()"
                    ))
                } else {
                    None
                }
            })
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states_module_name = states_module_name.as_str();
        let states = nodes
            .iter()
            .map(|node| {
                let resolved_name = node
                    .resolved_name
                    .clone()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    format!(".add_sub_state::<{states_module_name}::{resolved_name}>()")
                } else {
                    format!(".add_state::<{states_module_name}::{resolved_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    Ok(formatdoc! {"
        use bevy::prelude::AppExtStates;
        #[allow(missing_docs)]
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "})
}

pub(crate) fn remove_root_node(nodes: &mut Vec<NodeData>) {
    nodes.remove(0);
    for node in nodes {
        assert!(node.index > 0);
        node.index -= 1;
        assert!(node.depth > 0);
        node.depth -= 1;
        if let Some(parent_index) = &mut node.parent {
            node.parent = (*parent_index > 0).then(|| *parent_index - 1);
        }
    }
}

pub(crate) fn process_parse_nodes(
    parse_nodes: Vec<ParseNode<'_>>,
    naming_scheme: NamingScheme,
    root_state_name: Option<String>,
) -> Result<Vec<NodeData>, ProcessingError> {
    // add the implicit root_node according to config
    let root_node = if let Some(root_state_name) = &root_state_name {
        ParseNode::Enum(root_state_name.as_str().into(), parse_nodes)
    } else {
        // add a temporary root node
        ParseNode::List("".into(), parse_nodes)
    };
    let mut nodes = flatten_root_parse_node(root_node);
    // remove temporary root node
    if root_state_name.is_none() {
        remove_root_node(&mut nodes);
    }
    apply_naming_scheme(naming_scheme, &mut nodes)?;
    Ok(nodes)
}

pub fn convert_parse_nodes_into_plugin_source(
    parse_nodes: Vec<ParseNode<'_>>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let nodes = process_parse_nodes(
        parse_nodes,
        config.naming_scheme,
        config.root_state_name.clone(),
    )?;
    assert!(nodes.iter().all(|node| node.resolved_name.is_some()));
    build_plugin_source(nodes, config)
}
