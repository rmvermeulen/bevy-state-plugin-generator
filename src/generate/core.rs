use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::parsing::parse_config;
use crate::prelude::PluginConfig;
use crate::processing::{ProcessingError, convert_nodes_into_plugin_source};

pub(crate) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub(crate) fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub(crate) fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> Result<String, String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source.as_ref())
        .stderr_to_stdout()
        .read()
        .map_err(|e| format!("rustfmt: {e:?}"))
        .map(|source| {
            if source.ends_with(|c: char| c.is_newline()) {
                source
            } else {
                source + "\n"
            }
        })
}

pub(crate) fn generate_state_plugin_source(
    input_source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, ProcessingError> {
    let (unparsed, nodes) = parse_config(input_source)?;
    let mut output = convert_nodes_into_plugin_source(nodes, plugin_config)?;

    // if we're writing to a file we add a header with some information
    output = if let Some(src_path) = src_path {
        if !unparsed.trim().is_empty() {
            return Err(ProcessingError::InvalidConfig(unparsed.to_string()));
        }
        let debug_info = generate_debug_info(src_path, input_source);
        [debug_info, output].join("\n")
    } else {
        [unparsed, &output].join("\n")
    };

    format_source(&output).map_err(ProcessingError::Formatting)
}
