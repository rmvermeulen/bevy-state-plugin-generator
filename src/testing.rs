pub use insta::*;
pub use rstest::*;
pub use speculoos::prelude::*;

pub use crate::set_snapshot_suffix;

#[macro_export]
#[allow(missing_docs)]
macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

pub mod node_data {
    use bevy_utils::default;

    use super::*;
    use crate::processing::NodeData;

    fn node<S: ToString>(name: S, index: usize, depth: usize, parent: Option<usize>) -> NodeData {
        let name = name.to_string();
        NodeData {
            index,
            parent,
            resolved_name: Some(name.clone()),
            name,
            depth,
            ..default()
        }
    }

    #[fixture]
    pub fn single_node() -> Vec<NodeData> {
        vec![node("RootState", 0, 0, None)]
    }

    #[fixture]
    pub fn nested_example() -> Vec<NodeData> {
        vec![
            node("RootState", 0, 0, None),
            node("Menu", 1, 1, Some(0)),
            node("Options", 2, 2, Some(1)),
            node("Audio", 3, 3, Some(2)),
            node("Video", 4, 3, Some(2)),
        ]
    }
}

pub mod parse_node {
    use super::*;
    use crate::parsing::Node;

    #[fixture]
    pub fn duplicate_names() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("A")])
    }

    #[fixture]
    pub fn overlapping_names() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::enumeration("Root", [Node::singleton("A")]),
                Node::singleton("A"),
            ],
        )
    }

    #[fixture]
    pub fn enum_root_a() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn enum_root_ab() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn enum_root_a_b() -> Node<'static> {
        Node::enumeration("Root", [Node::enumeration("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn enum_root_a_b_up_c() -> Node<'static> {
        Node::enumeration(
            "Root",
            [
                Node::enumeration("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn list_root_a() -> Node<'static> {
        Node::list("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn list_root_ab() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn list_root_a_b() -> Node<'static> {
        Node::list("Root", [Node::list("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn list_root_a_b_up_c() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::list("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn nested_example() -> Node<'static> {
        Node::enumeration(
            "Menu",
            [
                Node::singleton("Main"),
                Node::comment("these are the options"),
                Node::enumeration(
                    "Options",
                    [
                        Node::singleton("Graphics"),
                        Node::singleton("Audio"),
                        Node::singleton("Gameplay"),
                    ],
                ),
                Node::enumeration("Game", [Node::singleton("Save"), Node::singleton("Load")]),
            ],
        )
    }
}
