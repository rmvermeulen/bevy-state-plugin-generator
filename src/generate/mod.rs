pub(crate) mod core;
#[cfg(test)]
mod tests;

use std::path::Path;
use std::{fs, io};

use itertools::{Itertools, concat};

use crate::generate::core::generate_state_plugin_source;
use crate::parsing::header::parse_template_header;
use crate::prelude::PluginConfig;
use crate::processing::ProcessingError;

/// The kinds of errors that can occur
#[derive(Debug, thiserror::Error)]
pub enum GeneratorError {
    /// A fs-related error occurred
    #[error("Io Error: {0}")]
    Io(#[from] io::Error),
    /// The content is wrong
    #[error("Processing Error: {0}")]
    Processing(#[from] ProcessingError),
}

/// ```rust no_run
/// use bevy_state_plugin_generator::prelude::*;
/// fn main() {
///   update_template(
///     "src/generated_states.rs",
///     PluginConfig::default()
///   ).expect("Failed to generate plugin!");
/// }
/// ```
#[cfg_attr(coverage_nightly, coverage(off))]
pub fn update_template(
    template_path: impl AsRef<Path>,
    mut plugin_config: PluginConfig,
) -> Result<(), GeneratorError> {
    let src_display = template_path.as_ref().to_string_lossy();
    println!("cargo:rerun-if-changed={src_display}");
    let source = std::fs::read_to_string(&template_path)?;
    let header = parse_template_header(&source, &mut plugin_config);

    let plugin_source =
        generate_state_plugin_source(&header.template.join("\n"), plugin_config, None)?;

    let header = concat([
        header
            .info_block
            .into_iter()
            .map(|line| format!("// {line}"))
            .collect_vec(),
        // vec![
        //     format!("comment_block={comment_block:?}"),
        //     format!("info_and_warnings={info_and_warnings:?}"),
        // ],
        header
            .comments_block
            .into_iter()
            .map(String::from)
            .collect_vec(),
    ])
    .join("\n");

    fs::write(&template_path, format!("{header}\n\n{plugin_source}")).map_err(Into::into)
}

/// ```rust no_run
/// use bevy_state_plugin_generator::prelude::*;
/// fn main() {
///   generate_plugin(
///     "src/states.txt",
///     "src/generated_states.rs",
///     PluginConfig::default()
///   ).expect("Failed to generate plugin!");
/// }
/// ```
#[cfg_attr(coverage_nightly, coverage(off))]
pub fn generate_plugin(
    src: impl AsRef<Path>,
    dst: impl AsRef<Path>,
    plugin_config: PluginConfig,
) -> Result<(), GeneratorError> {
    let src_display = src.as_ref().to_string_lossy();
    println!("cargo:rerun-if-changed={src_display}");
    let source = std::fs::read_to_string(&src)?;
    let source = generate_state_plugin_source(&source, plugin_config, Some(&src_display))?;
    fs::write(dst, source).map_err(Into::into)
}
