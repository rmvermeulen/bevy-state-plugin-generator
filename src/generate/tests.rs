use std::time::Duration;

use insta::{assert_debug_snapshot, assert_snapshot};
use itertools::Itertools;
use rstest::{Context, rstest};

use crate::generate::{format_source, generate_debug_info, generate_state_plugin_source};
use crate::parsing::ParseNode;
use crate::processing::{apply_naming_scheme, flatten_parse_node, try_parse_node_into_final_source};
use crate::testing::parse_node;
use crate::{NamingScheme, PluginConfig, set_snapshot_suffix};

#[cfg(feature = "rustfmt")]
const RUSTFMT: &str = "_rustfmt";
#[cfg(not(feature = "rustfmt"))]
const RUSTFMT: &str = "_no_rustfmt";

#[rstest]
#[timeout(Duration::from_millis(250))]
#[async_std::test]
async fn test_format_source() {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("formatted{suffix}");
    let formatted = format_source("fn main(){println!(\"Hello, world!\");}");
    assert_snapshot!(formatted);
}

#[rstest]
fn test_generate_states_plugin() {
    let root_state = ParseNode::enumeration(
        "GameState",
        [
            ParseNode::singleton("Loading"),
            ParseNode::enumeration(
                "Ready",
                [
                    ParseNode::enumeration(
                        "Menu",
                        [
                            ParseNode::singleton("Main"),
                            ParseNode::singleton("Options"),
                        ],
                    ),
                    ParseNode::enumeration(
                        "Game",
                        [
                            ParseNode::singleton("Playing"),
                            ParseNode::singleton("Paused"),
                            ParseNode::singleton("GameOver"),
                        ],
                    ),
                ],
            ),
        ],
    );
    assert_snapshot!(try_parse_node_into_final_source(root_state, Default::default()).unwrap());
}

#[rstest]
#[case("root.txt", "RootState")]
#[case("fruits.txt", "Apple Orange { O1 O2 }")]
fn test_generate_debug_info(#[case] src_path: &str, #[case] source: &str) {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}{suffix}");
    assert_snapshot!(generate_debug_info(src_path, source));
}

#[rstest]
#[case("root.txt", "RootState", PluginConfig::default())]
#[case(
    "mixed-nested-states.txt",
    "A { B [C[ } D { E F [ G H ] I }",
    PluginConfig::default()
)]
fn test_generate_full_source(
    #[case] src_path: &str,
    #[case] source: &str,
    #[case] plugin_config: PluginConfig,
) {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}{suffix}");
    assert_snapshot!(generate_state_plugin_source(source, plugin_config, Some(src_path)).unwrap());
}

#[rstest]
#[case("root.txt", "RootState", NamingScheme::Full)]
#[case("root.txt", "RootState", NamingScheme::Short)]
fn test_naming_scheme(
    #[case] src_path: &str,
    #[case] source: &str,
    #[case] naming_scheme: NamingScheme,
) {
    let suffix = cfg!(feature = "rustfmt")
        .then_some("_rustfmt")
        .unwrap_or_default();
    set_snapshot_suffix!("{src_path}_{naming_scheme}{suffix}");
    assert_snapshot!(
        generate_state_plugin_source(
            source,
            PluginConfig {
                naming_scheme,
                ..Default::default()
            },
            Some(src_path),
        )
        .unwrap()
    );
}

fn generate_all_type_definitions(node: ParseNode<'_>, naming_scheme: NamingScheme) -> Vec<String> {
    let mut nodes = flatten_parse_node(node);
    apply_naming_scheme(naming_scheme, &mut nodes).unwrap();
    nodes
        .into_iter()
        .map(|node| format!("{} -> {}", node.name, node.resolved_name.unwrap()))
        .collect_vec()
}

#[rstest]
#[case::enum_root_a(parse_node::enum_root_a())]
#[case::enum_root_ab(parse_node::enum_root_ab())]
#[case::enum_root_a_b(parse_node::enum_root_a_b())]
#[case::enum_root_a_b_up_c(parse_node::enum_root_a_b_up_c())]
#[case::list_root_a(parse_node::list_root_a())]
#[case::list_root_ab(parse_node::list_root_ab())]
#[case::list_root_a_b(parse_node::list_root_a_b())]
#[case::list_root_a_b_up_c(parse_node::list_root_a_b_up_c())]
#[case::nested_example(parse_node::nested_example())]
fn snapshots(
    #[context] context: Context,
    #[values(NamingScheme::Full, NamingScheme::Short, NamingScheme::None)]
    naming_scheme: NamingScheme,
    #[case] node: ParseNode,
) {
    set_snapshot_suffix!(
        "{}_{naming_scheme:?}{RUSTFMT}",
        context.description.unwrap()
    );
    assert_debug_snapshot!(generate_all_type_definitions(node, naming_scheme));
}
