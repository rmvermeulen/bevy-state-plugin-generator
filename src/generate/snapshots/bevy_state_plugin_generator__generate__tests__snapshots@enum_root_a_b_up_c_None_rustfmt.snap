---
source: src/generate/tests.rs
expression: "generate_all_type_definitions(node, naming_scheme)"
---
[
    "Root -> Root",
    "A -> A",
    "C -> C",
    "B -> B",
]
